src/core/or/circuitpadding.h

```diff
@@ -0,0 +1,539 @@
+/*
+ * Copyright (c) 2017, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+/**
+ * \file circuitpadding.h
+ * \brief Header file for circuitpadding.c.
+ **/
+#ifndef TOR_CIRCUITPADDING_H
+#define TOR_CIRCUITPADDING_H
+
+#include "circpad_negotiation.h"
+#include "lib/evloop/timers.h"
+
+typedef struct circuit_t circuit_t;
+typedef struct origin_circuit_t origin_circuit_t;
+typedef struct cell_t cell_t;
+
+/**
+ * Circpad state specifier.
+ *
+ * Each circuit has up to two state machines, and each state
+ * machine consists of these states. Machines transition between
+ * these states using the event transition specifiers below.
+ */
+typedef enum {
+  CIRCPAD_STATE_START = 0,
+  CIRCPAD_STATE_BURST = 1,
+  CIRCPAD_STATE_GAP = 2,
+  CIRCPAD_STATE_END = 3
+} circpad_statenum_t;
```

Comment:
I don't understand why we believe that 4 states is the right number.  Like, how sure are we there won't be more later?

Reply:
Since 2006 nobody has needed more than 2 states and 2 machines, for at least 4 different applications of this research concept and several independent implementations. However, it may be the case that people will invent more creative ways to use this system than have been studied previously, since it is now a real thing that can be used relatively easily.

Do you think we should make the number of states fully arbitrary? I could see a remote possibility of someone building up a far more complex HMM-style machine using way more states. I am just not sure if it is worth the complexity and potential clarity loss.

Reply:
Hm.  So, I guess 4 is probably good enough for now, especially if we don't think these need to be serialized except in torrc file?

Reply:
Thanks for suggesting this Nick. I'm also not sure whether this is worth doing, but I always thought that using just 4 states is really arbitrary.

I'm not sure if going HMM-style is the right way to do it.

But here is some pseudocode of generalizing the current 4 state system to an arbitrary number.

However, I don't really know how we could serialize this in the consensus, as Nick mentions.

```suggestion
typedef struct circpad_machine_event_t {
  int from_state;
  int to_state;
  enum circpad_event_t
};

typedef struct circpad_machine_state_t {
  char *name;
  circpad_machine_event_t *transition_events;
  circpad_state_t;
}

static int
initialize_machine_states(machine)
{
  machine_register_new_state(machine, "alpha");
  machine_register_new_state(machine, "beta");
  machine_register_new_state(machine, "gamma");

  /* Register transition events from START -> BURST */
  retval = machine_register_transition_events(machine, "alpha", "beta",
                                              CIRCPAD_EVENT_PADDING_RECV |
                                              CIRCPAD_EVENT_NONPADDING_RECV);
  if (retval < 0) { goto err; }

  /* Register transition events from BETTA -> GAMMA */
  retval = machine_register_transition_events(machine, "beta", "gamma", CIRCPAD_EVENT_NONPADDING_RECV);
  if (retval < 0) { goto err; }

  retval = machine_register_cancel_events(machine, "beta", CIRCPAD_EVENT_PADDING_RECV);
  if (retval < 0) { goto err; }
}
```

Reply:
Ok this was very simply to do so I just did it. 

36f85b04c2706d13a6b4086d39eacfab6345f0d2 has the header changes to support 255 states. 139ff5c6b0e78ff8edd54b1797088eee9e50b5c4 is the changes to the implementation. c41eefa615afd415857a7a8e8bbf0ff54fb4bf4d is the tests.

We don't need complicated registration APIs unless we want custom-defined string names. I don't think that is necessary. I think for custom states, we can just say "CUSTOM" or print the number.

With this change, the padding machines are now effectively a full hidden markov model. This means that future padding systems can do things like simulate entire browsing sessions across a variety of simulated pages (one or more pages per state). This is a very powerful ability, and worth providing.

Remaining question is if we should invert the state->event transition array to be event->state mapping, as nick suggested in https://github.com/torproject/tor/pull/461#discussion_r231697374. Now that we can have many more states than events, inverting may be a good idea. Also, if we invert, we can probably support up to 65535 states, which we maybe also want to do, because with that inversion, going from 255 to 65535 costs a byte and nothing else.

Reply:
Ok, I also expanded the max number of states to 65535 and inverted the transition array. See #461 (comment)